sdk/core/azure-json/src/main/java/com/azure/json/JsonWriteContext.java

```diff
@@ -0,0 +1,181 @@
+     * {@link JsonWriteContext#ROOT} and the operation is {@link JsonWriteOperation#END_OBJECT} or
+     * {@link JsonWriteOperation#END_ARRAY} and the parent context is {@link JsonContext#FIELD_VALUE} that will be
+     * completed as well as the value has completed writing.
+     * <p>
+     * Operations {@link JsonWriteOperation#START_OBJECT}, {@link JsonWriteOperation#START_ARRAY}, and {@link
+     * JsonWriteOperation#FIELD_NAME} create a child context where the current context becomes the parent context.
+     * <p>
+     * Lastly, {@link JsonWriteOperation#FIELD_AND_VALUE} is a special case as it's performing both open state and
+     * close state operations, so it's self-closing. This results in the current context being return.
+     *
+     * @param operation The {@link JsonWriteOperation} triggering the update.
+     * @return The updated writing context.
+     */
+    public JsonWriteContext updateContext(JsonWriteOperation operation) {
+        // Simple value has two scenarios:
+        //
+        // - Current context is the root, writing a simple value completes the JSON stream and the writing context
+        //   becomes COMPLETE.
+        // - Current context isn't the root, writing context becomes the parent context and the context is completed.
+        if (operation == JsonWriteOperation.SIMPLE_VALUE) {
+            return context == JsonContext.ROOT ? COMPLETED : parent;
+        }
+
+        // Ending an array or object has three scenarios, but before the scenarios play out the current context is
+        // completed. The scenarios are:
+        //
+        // - Parent context is the root, closing the array or object completes the JSON stream and the writing context
+        //   becomes complete.
+        // - Parent context is a FIELD_VALUE, closing the array or object completes the field value and the writing
+        //   context becomes the grandparent context.
+        // - The parent context is a wrapping array or object, return the parent context.
+        if (operation == JsonWriteOperation.END_ARRAY
+            || operation == JsonWriteOperation.END_OBJECT) {
+            JsonWriteContext toReturn = parent;
+
+            // Parent context is the root, complete writing by returning the COMPLETED context.
+            if (toReturn.context == JsonContext.ROOT) {
+                return COMPLETED;
+            }
+
+            // Parent context is a FIELD_VALUE, close the field context by returning the grandparent context.
+            if (toReturn.context == JsonContext.FIELD_VALUE) {
+                return toReturn.parent;
+            }
+
+            // Otherwise, just return the parent context.
+            return toReturn;
+        }
+
+        // The next set of checks are straight forward and return a new sub-context.
+        if (operation == JsonWriteOperation.START_OBJECT) {
+            return new JsonWriteContext(this, JsonContext.OBJECT);
```

Comment:
it seems we need a new context per token, is there a way to avoid allocations?

Reply:
Correct, we do create a new context per token as there is a parent-child structure. To avoid `JsonWriteContext` allocations this could be changed to being a stack. That would require some testing to determine which is better

Reply:
Additionally, `JsonWriteContext` could have more information packed into it in the future such as position within an array to help support potential optimizations if the underlying `JsonWriter` implementation supports it